File: Data Structures & Algorithms/burst-balloons/submission-4.java
class Solution {
    public int maxCoins(int[] nums) {
        int[] newNums = new int[nums.length + 2];
        newNums[0] = 1;
        newNums[nums.length + 1] = 1;

        for (int i = 0; i < nums.length; i++) {
            newNums[i + 1] = nums[i];
        }

        Map<String, Integer> memo = new HashMap<>();

        return maxCoinsHelper(newNums, 1, nums.length, memo);
    }

    private int maxCoinsHelper(int[] nums, int left, int right, Map<String, Integer> memo) {
        if (left > right) {
            return 0;
        }
        String key = left + "-" + right;

        if (memo.containsKey(key)) {
            return memo.get(key);
        }

        int max = 0;
        for (int i = left; i <= right; i++) {
            // iterate through each spot, and see if we find the max
            int currCoins = nums[left - 1] * nums[i] * nums[right + 1];
            max = Math.max(max, currCoins + maxCoinsHelper(nums, left, i - 1, memo) + maxCoinsHelper(nums, i + 1, right, memo));
        }

        memo.put(key, max);

        return max;
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    lengthOfLIS(nums) {
        // algorithm
        // create an array of length nums
        // that stores the length of the "chain" we can make starting at a given index and value
        // set the last value of the array to 1 (since we reached the end)
        // go from right to left, and at each nums[i], we look through the remaining values to the right
        // and see the longest chain we can prepend to.

        if (!nums) {
            return 0;
        }

        if (nums.length === 1) {
            return 1;
        }

        const memo = new Array(nums.length).fill(0);
        memo[nums.length - 1] = 1;

        let result = 1;
        for (let i = nums.length - 2; i >= 0; i--) {
            // go through all values in memo to the right of i (j), and if nums[j] > nums[i]
            // record the highest chain value we find and store that in memo[i]
            let longestChain = 0;
            for (let j = i + 1; j < nums.length; j++) {
                if (nums[j] > nums[i] && memo[j] > longestChain) {
                    longestChain = memo[j];
                }
            }
            memo[i] = 1 + longestChain;
            result = Math.max(result, memo[i]);
        }

        return result;
    }
}
